### PR TITLE
Added a cache with a read-write lock for components

### PR DESCRIPTION
**Pull request by Casper-NS**
This PR adds a cache for models (i.e. the Component hashmap).
This was added due to our gRPC changes, where components are sent together with each query.
To avoid parsing the components every time, we thought it would be smarter to cache them.

This is also in preparation for sharing components across threads for multithreading. Therefore the Hashmap is in a RWlock, which means we can have many readers, but only one writer at a time.

Components have been put into Arcs, to make them thread safe.

## Perfomance improvements, based on the test in the new benchmark branch #68 .
```
send_query_same_components
                        time:   [1.4232 ms 1.4423 ms 1.4590 ms]
                        change: [-63.223% -62.700% -62.155%] (p = 0.00 < 0.05)
                        Performance has improved.

send_query_different_components
                        time:   [3.6976 ms 3.7259 ms 3.7559 ms]
                        change: [-1.3595% -0.4077% +0.6106%] (p = 0.43 > 0.05)
                        No change in performance detected.
```

### CODE REVIEW

File: src/DataReader/component_loader.rs [hunk 1]
```diff
@@ -6,6 +6,27 @@ use crate::ModelObjects::queries::Query;
 use crate::ModelObjects::system_declarations::SystemDeclarations;
 use crate::System::input_enabler;
 use std::collections::HashMap;
+use std::sync::{RwLock, Arc};
+
+
+/// A struct used for caching the models sent with queries.
+#[derive(Debug, Default, Clone)]
+pub struct ModelCache {
+    Cache: Arc<RwLock<HashMap<String, Arc<ComponentContainer>>>>,
```
Comment (mads256h) [inline src/DataReader/component_loader.rs:15]: Cache -> cache

File: src/DataReader/component_loader.rs [hunk 2]
```diff
@@ -6,6 +6,27 @@ use crate::ModelObjects::queries::Query;
...
+}
+
+impl ModelCache {
+    pub fn get_model(&self, containerHash: &str) -> Arc<ComponentContainer> {
```
Comment (mads256h) [inline src/DataReader/component_loader.rs:19]: This should return an optional.
If the components are not in the cache we want to insert them later

File: src/DataReader/component_loader.rs (hunk 1 above)
Comment (mads256h) [inline src/DataReader/component_loader.rs:15]: The hashmap should be mapping from an u32 to a component container

File: src/DataReader/component_loader.rs [hunk 3]
```diff
@@ -6,6 +6,27 @@ use crate::ModelObjects::queries::Query;
...
+        match self.Cache.read().unwrap().get(containerHash) {
+            Some(model) => model.clone(),
+            None => panic!("The component container with '{}' could not be retrieved from the cache", containerHash) 
+        }
+    }
+
+    pub fn insert_model(&mut self, containerHash: String, container: ComponentContainer) {
+        self.Cache.write().unwrap().entry(containerHash).or_insert(Arc::new(container));
```
Comment (mads256h) [inline src/DataReader/component_loader.rs:27]: I would not use `entry` here. There should be an insert method directly on the hashmap.

**Review by mads256h — CHANGES_REQUESTED** [review]

File: src/DataReader/component_loader.rs (hunk 1 above)
Comment (Casper-NS) [reply src/DataReader/component_loader.rs]: done

File: src/DataReader/component_loader.rs (hunk 2 above)
Comment (Casper-NS) [reply src/DataReader/component_loader.rs]: done

File: src/DataReader/component_loader.rs (hunk 1 above)
Comment (Casper-NS) [reply src/DataReader/component_loader.rs]: done

File: src/DataReader/component_loader.rs [hunk 4]
```diff
@@ -14,7 +58,7 @@ pub trait ComponentLoader {
 
 #[derive(Debug, Default, Clone)]
 pub struct ComponentContainer {
-    pub loaded_components: HashMap<String, Component>,
+    pub loaded_components: Arc<HashMap<String, Component>>,
```
Comment (henneboy) [inline src/DataReader/component_loader.rs:61]: plz use components map

File: src/DataReader/component_loader.rs [hunk 5]
```diff
@@ -31,7 +75,7 @@ impl ComponentLoader for ComponentContainer {
 }
 
 impl ComponentContainer {
-    pub fn new(map: HashMap<String, Component>) -> Self {
+    pub fn new(map: Arc<HashMap<String, Component>>) -> Self {
```
Comment (henneboy) [inline src/DataReader/component_loader.rs:78]: plz use components map

**Review by henneboy — CHANGES_REQUESTED** [review]
Almost LGTM

**Review by mads256h — DISMISSED** [review]
lgtm

File: src/DataReader/component_loader.rs (hunk 5 above)
Comment (Casper-NS) [reply src/DataReader/component_loader.rs]: done

**Review by henneboy — APPROVED** [review]

**Review by S3JER — APPROVED** [review]
Ser fint ud

**Pull request merged**